agent/consul/rpc.go

```diff
@@ -203,6 +205,13 @@ func canRetry(args interface{}, err error) bool {
 		return true
 	}
 
+	// If we were chunking and the term changed, resubmit against the new
+	// leader
+	intErr, ok := args.(error)
+	if ok && intErr == raftchunking.ErrTermMismatch {
```

Comment:
Does this comparison def work through a real RPC call? I'm 80% sure it won't from experience. We have some issues where RPCs to the same instance go through in-memory transport which doesn't do a full msgpack encode/decode cycle so we have hit things like this that pass tests but don't actually work in real life due to type mangling. e.g. msgpack will decode the result into a plain old `error` value which won't compare equal to the specific error value in the library. We end up needing to string match the message in this case currently 😭 .

Reply:
It's a shame this is hard to test well. Testing this properly would be very timing sensitive and hard to make robust. It would be nice if we could test against a mock server but we don't have one currently that makes it easy to control the error return etc.

I think if we make it a string match then we can verify in smoke tests for now.

Reply:
FWIW I had the same thing here: https://github.com/hashicorp/consul/blob/3a4e38a13ee9245bad315f9f239cc49995c6536e/agent/cache-types/connect_ca_leaf.go#L543 initially I just compared with the public error type and it worked in tests where RPC was in-memory and types were preserved, but failed in integration tests once RPCs were going through msgpack and loosing type info.

Reply:
Fixed. I went one step further and did a strings.Contains() so that if in the future there is any errwrap that happens (or usage of the newer error stuff coming in Go 1.13) then it will still work.